kuzzleio/sdk-go
# C++11 portage

## What does this PR do ?

Now use C++11.

Implemented features:
- `lambda` for listeners + swig wrapping of those lambda
- use of `foreach`
- use of `nullptr`
- use of `override`
- use of `noexcept`

Note:
- I removed all use of `throw` as it is deprecated in C++11 and removed in C++14
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2010/n3051.html
- Since I removed the C++ throw directive I also had to change the wrapping way of exceptions for JAVA


### How should this be manually tested?

- Make sure it builds

```sh
cd internal/wrappers
make cpp java
```

- Check if `internal/wrappers/build/java/build/io/kuzzle/sdk/NotificationListener.java` have `onMessage` abstract
- Check if `internal/wrappers/build/java/build/io/kuzzle/sdk/EventListener.java` have `trigger` abstract
- C++ snippet:
```cpp
#include "kuzzle.hpp"
#include "realtime.hpp"
#include "document.hpp"
#include <functional>
#include <iostream>
#include <unistd.h>

using namespace kuzzleio;

int main() {
  try {
    Kuzzle* kuzzle = new Kuzzle("localhost", nullptr);
    kuzzle->connect();

    notification_result* result;

    std::function<void(const kuzzleio::notification_result*)> listener = [&result](const notification_result* res) {
      result = const_cast<notification_result*>(res);
    };
    kuzzle->realtime->subscribe("index", "foo", "{}", &listener);
    kuzzle->document->create("index", "foo", "", "{\"foo\":\"bar\"}");
    sleep(1);
    std::cout << result->result->content << std::endl;

  } catch (KuzzleException& e) {
    std::cerr << e.getMessage() << std::endl;
  }
}
```
